File: app/src/main/java/com/example/uscatterbrain/network/CappedInputStream.java
package com.example.uscatterbrain.network;

import java.io.IOException;
import java.io.InputStream;

/**
 * Since protobuf's api for reading streams into bytebuffers lacks
 * the ability to read partial streams, this hack should
 * trick it into not devouring the entire inputstream
 */
public class CappedInputStream extends InputStream {
    private InputStream mStream;
    private int mRemaining;
    public CappedInputStream(InputStream stream, int cap) {
        super();
        mRemaining = cap;
        mStream = stream;
    }

    @Override
    public int read(byte[] b) throws IOException {
        if (mRemaining <= 0)
            return -1;

        int bytesread = mStream.read(b, 0, mRemaining);
        mRemaining -= bytesread;
        return bytesread;
    }

    @Override
    public int read(byte[] b, int off, int len) throws IOException {
        if (mRemaining <= 0)
            return -1;
        int l = mRemaining;

        if (mRemaining + off > b.length)
            l = b.length - off;

        int bytesread = mStream.read(b, off, l);
        mRemaining -= bytesread;
        return  bytesread;
    }

    @Override
    public long skip(long n) throws IOException {
        return mStream.skip(n);
    }

    @Override
    public int available() throws IOException {
        return Math.min(mStream.available(), mRemaining);
    }

    @Override
    public void close() throws IOException {
        mStream.close();
    }

    @Override
    public synchronized void mark(int readlimit) {
        mStream.mark(readlimit);
    }

    @Override
    public synchronized void reset() throws IOException {
        mStream.reset();
    }

    @Override
    public boolean markSupported() {
        return mStream.markSupported();
    }

    @Override
    public int read() throws IOException {
        if (mRemaining > 0) {
            int val = mStream.read();
            mRemaining--;
            return val;
        } else {
            return -1;
        }
    }
}

File: app/src/main/java/net/ballmerlabs/uscatterbrain/network/CappedInputStream.kt
package net.ballmerlabs.uscatterbrain.network

import java.io.IOException
import java.io.InputStream

/**
 * Since protobuf's api for reading streams into bytebuffers lacks
 * the ability to read partial streams, this hack should
 * trick it into not devouring the entire inputstream
 */
class CappedInputStream(private val mStream: InputStream, private var mRemaining: Int) : InputStream() {
    @Throws(IOException::class)
    override fun read(b: ByteArray): Int {
        if (mRemaining <= 0) return -1
        val bytesread = mStream.read(b, 0, mRemaining)
        mRemaining -= bytesread
        return bytesread
    }

    @Throws(IOException::class)
    override fun read(b: ByteArray, off: Int, len: Int): Int {
        if (mRemaining <= 0) return -1
        var l = mRemaining
        if (mRemaining + off > b.size) l = b.size - off
        val bytesread = mStream.read(b, off, l)
        mRemaining -= bytesread
        return bytesread
    }

    @Throws(IOException::class)
    override fun skip(n: Long): Long {
        return mStream.skip(n)
    }

    @Throws(IOException::class)
    override fun available(): Int {
        return Math.min(mStream.available(), mRemaining)
    }

    @Throws(IOException::class)
    override fun close() {
        mStream.close()
    }

    @Synchronized
    override fun mark(readlimit: Int) {
        mStream.mark(readlimit)
    }

    @Synchronized
    @Throws(IOException::class)
    override fun reset() {
        mStream.reset()
    }

    override fun markSupported(): Boolean {
        return mStream.markSupported()
    }

    @Throws(IOException::class)
    override fun read(): Int {
        return if (mRemaining > 0) {
            val `val` = mStream.read()
            mRemaining--
            `val`
        } else {
            -1
        }
    }

}